### PR TITLE
Rework the way we override Docker Compose settings

### PR DESCRIPTION
**Pull request by jonathansharman**
### Dependencies

- Uses https://github.com/nicheinc/development/pull/327

### Documentation

- Related [`auth` CI task](https://app.asana.com/0/1202563967140527/1203329473602218/f)
- Related [slack thread](https://nicheinc.slack.com/archives/C0GN6575G/p1672947526051209)

### Description

This reworks the way that we use `docker-compose.ci-override.yml`, ultimately so that we can skip internal DNS servers in CI.

Currently, we create a `docker-compose.override.yaml` on the fly that sets the relevant service images to the images produced in the CI run, and then we merge `development/docker-compose.ci-override.yml` into it with `yq`. Unfortunately, we can't use this approach to _overwrite_ the list of DNS IPs - we can only append to the existing list. (See the note on the `dns` field in the [Docker Compose docs](https://docs.docker.com/compose/extends/#adding-and-overriding-configuration).)

With this change, we will instead overwrite the images in `docker-compose.ci-override.yml` with the CI images and then merge that file directly into `docker-compose.yml` with `yq`. Unlike Docker Compose, `yq`'s merge operation actually overwrites fields.

The observable effects of this change are:
1. The hard-coded CI overrides in the workflow will take precedence over `docker-compose.ci-override.yml` settings.
2. The final versions of all `docker-compose` fields will now contain only the values in `docker-compose.ci-override`, rather than situationally concatenating the two lists of values.

The first is just a side effect of my implementation and shouldn't matter unless for some reason we wanted to use `docker-compose.ci-override.yml` to change which service images are used. The second will allow us to actually remove unwanted DNS addresses in CI.

### Testing Considerations

The benefits of this change can be observed in [this `auth` run](https://github.com/nicheinc/auth/actions/runs/3875871196/jobs/6609355181#step:16:376), which uses this branch. In that run, only 3 test suites fail, down from [9 failed suites](https://github.com/nicheinc/auth/actions/runs/3833046534/jobs/6524088183#step:16:765) when using `v0`.

After the associated `development` branch is merged and this branch updated to point to `development@master` again (see below), we should rerun the workflow in `auth` and make sure nothing changes (should still have just 3 test suites failing and no other errors).

### Deployment 

~~This branch of the workflow currently checks out `development@x/override-dns`. Once https://github.com/nicheinc/development/pull/327 is merged, this branch should be updated to once again pull straight from `development@master`. Once that's done,~~ this can be merged straight into the `v0` branch.

### Versioning

This is sort of a patch. We should stay on `v0` by simply merging.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203802840423813

### CODE REVIEW

**Comment by bclarkx2** [comment]
This only work because the `yq` merge is changing the value of the top-level field `x-dns` field in `docker-compose.yml`, right? Since YAML anchors only have meaning inside the file they are defined in, I can't imagine the lines added in https://github.com/nicheinc/development/pull/327 do anything on their own.

```bash
[development]~/code/development↑{origin}(x/override-dns*).
$ docker-compose -f docker-compose.yml -f docker-compose.ci-override.yml config | yq '.services.auth.dns'
- 192.168.0.252
- 192.168.0.90
- 8.8.8.8
```

It's a little unintuitive to me that `docker-compose.ci-override.yml` would contain this `x-dns` override that actually doesn't do anything if you just combine them with the `docker-compose` override utilities.

But, I have no better ideas to suggest at the moment, though :sweat_smile:! Obviously it would be best if we could convince the services to truly inherit the DNS settings of the host they are running on, though I know we stalled out on that a year ago.

**Review by angelowilliams — APPROVED** [review]
LGTM! It doesn't feel amazing to be relying upon whatever logic `yq` uses to merge files, but in this case the solution is simple and reliable, so I'm on board. We can cross the bridge of what to do whenever we want to also have the normal `docker-compose.override.yml` behavior if that situation arises.

File: .github/workflows/action.yaml [hunk 1]
```diff
@@ -302,13 +303,12 @@ jobs:
           ENCODED_CREDENTIALS: ${{ secrets.DEV_AWS_CREDENTIALS }}
         shell: bash
         run: |
-          yq -n ".version = \"3.7\"" > docker-compose.override.yaml
           for service in ${{ steps.get-services.outputs.services }}; do
-            yq -i ".services.${service}.image = \"${service}:ci\"" docker-compose.override.yaml
+            yq -i ".services.${service}.image = \"${service}:ci\"" ${{ github.workspace }}/docker-compose.ci-override.yml
```
Comment (wmarshall) [inline .github/workflows/action.yaml:307]: I know we're pretty bought-in to `yq` at this point in time, but if we ever wanted to switch to good old `jq`, [this command exists](https://docs.docker.com/engine/reference/commandline/compose_convert/) and could be used like `docker compose convert --format json` (or `docker-compose config --format json`) to get you a happy resolved json file that might be easier to work with.

Comment (jonathansharman) [reply .github/workflows/action.yaml:307]: Good to know, thank you!

**Pull request merged**